### ISSUE TITLE
Negative-case type narrowing for class patterns when arguments are present or class is generic

### ISSUE THREAD

**Issue by to-ph** [microsoft/pyright#2970]
**Is your feature request related to a problem? Please describe.**

Yes, it seems that pyright does not support `TypeAlias` when using python 3.10's match statement.
For example, using this code snippet:

```python
@dataclass
class Ok(Generic[T]):
    val: T

@dataclass
class Err(Generic[E]):
    val: E

Result: TypeAlias = Ok[T] | Err[E]

def divide(a: int, b: int) -> Result[float, str]:
    ...  # Err if dividing by 0, Ok otherwise
    
match divide(1, 2):
  case Ok(val):
      ...  # val is float
  case Err(message):
      ...  # message is str
```

pyright outputs:
```
Cases within match statement do not exhaustively handle all values 
Unhandled type: "Result[float, str]"
```
when it should be smart enough to know that this was an exhaustive check

**Describe the solution you'd like**
The example above should type check without any complaints. The matching is exhaustive.

**Comment by to-ph** [comment]
Ah, I just attempted the above with the type signature of `divide` as:
```python
def divide(a: int, b: int) -> Ok[float] | Err[str]:
    ...
```
But pyright still complains the match statement is **not** exhaustive, so perhaps this isn't a limitation of TypeAlias

**Comment by erictraut** [comment]
This doesn't have anything to do with type aliases. Pyright doesn't currently support type narrowing in the negative (not matched) case for class patterns that have arguments or for generic classes. That means exhaustion isn't possible in the above example without providing a wildcard pattern case at the end of the match statement.

I need to think about this more, but I think it's possible to loosen these restrictions and support negative-case type narrowing in cases where the argument list does not constrain the pattern.

**Comment by to-ph** [comment]
Thanks for the speedy reply!
I'll continue to use the `assert_never` idiom you suggested in a separate issue. Otherwise this particular issue is not a problem for me, just a nice to have 😄

**Comment by erictraut** [comment]
The `assert_never` unfortunately won't work in this particular case because it relies on negative-case type narrowing.

**Comment by to-ph** [comment]
Yeah I've realized 😅, I just adapted it to
```python
def assert_never(x: Any) -> NoReturn:
    assert False, "Unhandled type: {}".format(type(x).__name__)
```
to make it work for these cases

**Comment by erictraut** [comment]
I've added support for negative narrowing based on arguments. The new logic also supports pattern classes that are generic. With this change, your above sample is correctly treated as exhaustive. This will be included in the next release.

**Comment by to-ph** [comment]
Thanks so much! I appreciate it 😄

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.218, which I just published. It will also be included in the next release of pylance.
